python/ray/train/callbacks/results_prepocessors/preprocessor.py

```diff
@@ -0,0 +1,44 @@
+import abc
+from typing import List, Dict
+
+
+class ResultsPreprocessor(abc.ABC):
+    """Abstract class for preprocessing Train results."""
+
+    @abc.abstractmethod
+    def preprocess(self, results: List[Dict]) -> List[Dict]:
+        """
+
+        Args:
+            results (List[Dict]): List of results from the training
+                function. Each value in the list corresponds to the output of
+                the training function from each worker.
+
+        Returns: A list of dictionaries. Each item in the list does *not*
+            need to correspond to a single worker, and it is expected for
+            the corresponding caller to understand the semantics of the
+            preprocessed results.
+
+        """
+        return results
+
+
+class SequentialResultsPreprocessor():
```

Comment:
Technically yes, but I found this to be cleaner and more extensible. Preprocessing is captured as a single `preprocess` step which can be customized as needed.

Reply:
Isn't the main purpose of `SequentialPreprocessor` to contain multiple preprocessors and enforce an ordering? These properties are already captured by any ordered iterable.

The main reasoning for using a list/iterable instead of a wrapper class is for usability. IMO, as a user, if I want to pass in multiple preprocessors to my callback, it would be much easier to just pass in a list instead of wrapping in a `SequentialResultsPreprocessor`.

I think the only change that would have to be made to support a list would be the `process_results` method in the base `TrainingCallback`.

But I'm curious what you had in mind for the extensibility use case?

Reply:
Yes that is correct!

The difference is  whether we want sequential preprocessing to be a first-class idea in `process_results` API. In my mind it is a secondary utility.

For extensibility:
1. If we have system level preprocessing (e.g. to support required preprocessing), I think it is cleaner to have the structure as `SequentialResultsPreprocessor([SystemPreprocessor, UserPreprocessor])` vs. concatenating 2 flattened lists.
2. I haven't thought this one through yet, but if we want to split preprocessing into different paths and then merge the results together, the Iterable API may not make sense.

Reply:
Ok I'm fine with keeping this as an internal concept, but would like to not expose this to users.

What do you think about still allowing users to pass in an iterable as the result preprocessors? I think that would just be adding these 2 lines to `process_results`

```python
if isinstance(results_preprocessors, Iterable):
    results_preprocessors = SequentialResultsPreprocessor(results_preprocessors)
```

Reply:
Personally I still don't think having the user to add `SequentialResultsPreprocessor(...)` is a significant overhead for the explicit indicator that they will be processed sequentially, and don't think it needs to be privatized.

There are a few of things I could nitpick about using the `Iterable`, but I'm wondering if we could postpone this until a clear use-case arises? The snippet you shared could easily be added if there's a clear indicator that this improves usability, whereas removal is usually less preferred.

Reply:
Commented this below, but to me the cleanest solution would be that training callbacks accept an optional list of result preprocessors in `__init__`. If this is provided, it overrides the default preprocessors that are set. Internally we can just convert this into a SequentialResultsPreprocessor. All callbacks should always be calling `super` anyways.

The benefit here is that this behavior is clearly documented in the docstring, and passing in args via a parameter is cleaner than having to set an attribute (which would be a pain to maintain backwards compatibility for).

Reply:
My concern is that it requires the user to define the entire `ResultsPreprocessor` (single or multiple) in the constructor vs. flattened args that are more understandable to the user (e.g. `workers_to_log`).

This goes back to the question of whether `ResultsPreprocessor`s should be part of the public API or developer API.

Reply:
We could accept both, but I agree if we want to have this be a developer API then it's fine to leave as is.